Add .NET peer implementation with TCP transport and GossipSub

leny62 opened the pull request:
## Overview
This PR adds a .NET peer implementation to the universal connectivity demo, which previously showcased connectivity between Go, Rust, and JavaScript. The new implementation demonstrates how libp2p can be used in the .NET ecosystem, providing a way for C# developers to participate in the same libp2p network.

## Features
- **Full-featured .NET peer** using the Nethermind.Libp2p library
- **TCP transport** protocol support
- **GossipSub** protocol for message publishing/subscription
- **Beautiful console UI** with separate panels for chat, peers, and system logs
- **Cross-implementation compatibility** with existing Go, and probably Rust (testing soon)

## Changes
- Added `dotnet-peer` directory with a complete implementation
- Created Console UI:
  - Terminal.Gui-based UI 
- Comprehensive README documentation for .NET peer setup and usage
- Added appropriate .gitignore rules for .NET projects

## Testing
The implementation has been tested for interoperability with:
- Go peer over TCP

## Live Demo
https://drive.google.com/file/d/1crse0KP1bfbFLx-_BkaG77uPYZ-mDI1G/view?usp=drive_link


## Protocol Support Matrix (Updated)
| Package         | Description                     | WebTransport | WebRTC | WebRTC-direct | QUIC | TCP |
| --------------- | ------------------------------- | ------------ | ------ | ------------- | ---- | --- |
| `js-peer`       | Browser Chat Peer in TypeScript | ✅           | ✅     | ✅            | ❌   | ❌  |
| `go-peer`       | Chat peer implemented in Go     | ✅           | ❌     | ✅            | ✅   | ✅  |
| `rust-peer`     | Chat peer implemented in Rust   | ❌           | ❌     | ✅            | ✅   | ❌  |
| `dotnet-peer`   | Chat peer implemented in .NET   | ❌           | ❌     | ❌            | 🚧   | ✅  |

This implementation expands the universal connectivity demo to include the .NET ecosystem, further demonstrating libp2p's cross-language interoperability capabilities.

dhuseby commented:
It looks like you hacked up the go client on the left for your video. Is that correct? It looks from the log output that you hard coded it to connect to your dotnet-peer. Then you must have changed the chat window code to show the full object received instead of just the chat message. That's no big deal, it's just not what users should expect.

Also, is there a way to re-direct all of the log tracing to the "Logs" tab in your UI instead of having it take over your terminal briefly? In Rust there's an easy way to do that: https://github.com/libp2p/universal-connectivity/pull/245/files#diff-4a6d015b4db92c87150fef3ceca232743a6bd871ee368ced697e338e81b2fb41

I'm not sure if there is one in DotNet. I'm pretty sure you just have to either inject code into the logging system that libraries/apps use...if not that, then you need to redirect stderr to a stream you read from and output to the Log tab.

Overally, this is excellent!! Great work. 🍻

dhuseby commented:
You don't have to hard code the go-peer to connect to your peer. It has a command line argument "-connect" that you can pass the multiaddr to (e.g. /ip4/127.0.0.1/tcp/9096/p2p/\<peer id\>).

leny62 commented:
Thank you for the feedback @dhuseby. I've made several improvements to address your comments:

## Updates in this revision:

1. **UI Display Fix**: 
   - Fixed the Terminal.GUI display issue by properly separating threading concerns
   - The chat UI now displays correctly within the console window, with all components properly organized
   - The application now correctly shows the welcome messages and connection information in the UI

2. **Log Redirection**: 
   - Implemented proper log redirection to the "Logs" tab in the UI
   - Added a custom logging provider that captures all logs (including library logs) and displays them in the UI
   - Console output is significantly cleaner, with most messages now appearing in the Logs tab

3. **Clean Architecture**:
   - Restructured the codebase into a clean architecture with proper separation of concerns
   - Implemented Domain, Application, Infrastructure, and Presentation layers
   - All components now interact through well-defined interfaces

4. **Demo Video**:
   - [Link
](https://drive.google.com/file/d/1-HpnSetB49fKxO1PxayCq-mt_6p_kM2Y/view?usp=drive_link)

flcl42 commented on dotnet-peer/Libp2pChat/Infrastructure/Logging/TerminalLogger.cs:
```diff
@@ -0,0 +1,100 @@
+public class TerminalLogger : IAppLogger
+{
+    private readonly IChatUI _chatUI;
+    private readonly ILogger _logger;
```
Redundant field and below

flcl42 commented on dotnet-peer/Libp2pChat/Infrastructure/Logging/TerminalLogger.cs:
```diff
@@ -0,0 +1,100 @@
+    /// <inheritdoc />
+    public void LogWarning(string message)
+    {
+        string formattedMessage = $"[{DateTime.Now:HH:mm:ss.fff}] [Warning] {message}";
```
The body can be a separate function that logs using the same formatting, to reduce duplication

flcl42 commented on dotnet-peer/Libp2pChat/Infrastructure/Libp2p/Libp2pChatService.cs:
```diff
@@ -0,0 +1,342 @@
+            
+            // Check if the message is JSON or plain text
+            bool isJsonMessage = raw.StartsWith("{");
+            if (isJsonMessage)
```
Not a good check, you can omit it and catch potential error in existing catch block below

flcl42 commented on dotnet-peer/Libp2pChat/Infrastructure/Libp2p/Libp2pChatService.cs:
```diff
@@ -0,0 +1,342 @@
+                var plainTextMessage = new ChatMessage(raw, "unknown", "peer");
+                
+                // Add Go peer to known peers
+                var goPeer = Peer.CreateGoPeer();
```
It can be any peer, go client is not a the only we expect to connect with. I'd suggest to change naming to something like peer or remotePeer

flcl42 commented on dotnet-peer/Libp2pChat/Infrastructure/Libp2p/Libp2pChatService.cs:
```diff
@@ -0,0 +1,342 @@
+            _localPeer = _peerFactory.Create(_identity);
+            
+            // Create multiaddresses to listen on
+            var localAddress = Multiaddress.Decode($"/ip4/127.0.0.1/tcp/9096/p2p/{PeerId}");
```
Is 9096 a fixed port? Can it be parametrized? Or random. To listen to random one you can pass 0 instead

flcl42 commented on dotnet-peer/Libp2pChat/Infrastructure/Libp2p/Libp2pChatService.cs:
```diff
@@ -0,0 +1,342 @@
+            
+            // Create multiaddresses to listen on
+            var localAddress = Multiaddress.Decode($"/ip4/127.0.0.1/tcp/9096/p2p/{PeerId}");
+            var allIfacesAddress = Multiaddress.Decode($"/ip4/{_listenAddress}/tcp/9096/p2p/{PeerId}");
```
"allIfacesAddress" includes local one, so no need to listen to it separately

dhuseby replied:
@flcl42 thanks for the review. @leny62 if you clean these up, I'll land this.

leny62 replied:
@flcl42 thanks for the feedback, on it.

flcl42 reviewed (changes requested):
There are 
`dotnet-libp2p-chat` - which seems to be redundant and broken
`Libp2pChat.csproj` - which seems to be working

Could you clean up the code form parts not needed. And also and solution file?

flcl42 commented on dotnet-peer/README.md:
````diff
@@ -0,0 +1,172 @@
+Clone the repository and navigate to the dotnet-peer directory:
+
+```bash
+git clone https://github.com/leny62/universal-connectivity
````
Can be changed to the main repo

leny62 replied:
@flcl42 , I'll address these changes ASAP.

@dhuseby , had created this branch `dotnet-peer-development` to merge the .NET to it until its production ready.

flcl42 commented on node-js-peer/package.json:
```diff
@@ -0,0 +1,41 @@
+{
```
This update is not required I suppose

leny62 replied:
@flcl42 , I think changes from other peers might be because I synced my folked repo.

flcl42 replied:
@dhuseby could you merge main into dotnet-peer-development please?

flcl42 reviewed (commented):
I'd also suggest to use ILogger interface instead of IAppLogger if possible. 
The pr also contains js peer, not sure if it's intentional.
It also created to be merged not into the main branch.
